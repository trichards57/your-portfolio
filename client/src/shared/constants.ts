export const ServerAudience = "https://tr-toolbox.me.uk/your-portfolio";
