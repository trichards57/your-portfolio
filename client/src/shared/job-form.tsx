import {
  Button,
  FormControl,
  FormControlLabel,
  Grid,
  InputLabel,
  Select,
  Switch,
  TextField,
} from "@material-ui/core";
import React from "react";
import { Save as SaveIcon } from "@material-ui/icons";
import { Skeleton } from "@material-ui/lab";
import { Outcome } from "../model/job";
import useSharedStyles from "./shared-styles";

interface JobFormProps {
  age: number | undefined;
  blueLights: boolean;
  canSubmit: boolean;
  category: number;
  drove: boolean;
  gender: "Male" | "Female" | undefined;
  isLoading: boolean;
  notes: string;
  outcome: Outcome;
  reflectionFlag: boolean;
  setAge(value: number | undefined): void;
  setBlueLights(value: boolean): void;
  setCategory(value: number): void;
  setDrove(value: boolean): void;
  setGender(value: "Male" | "Female" | undefined): void;
  setNotes(value: string): void;
  setOutcome(value: Outcome): void;
  setReflectionFlag(value: boolean): void;
  submit(): void;
}

export default function JobForm(props: JobFormProps) {
  const {
    age,
    blueLights,
    canSubmit,
    category,
    drove,
    gender,
    isLoading,
    notes,
    outcome,
    reflectionFlag,
    setAge,
    setBlueLights,
    setCategory,
    setDrove,
    setGender,
    setNotes,
    setOutcome,
    setReflectionFlag,
    submit,
  } = props;

  const sharedClasses = useSharedStyles();

  return (
    <form noValidate>
      <Grid container spacing={2}>
        <Grid item sm={12} md={2}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <TextField
              label="Age"
              type="number"
              id="age"
              className={sharedClasses.formItem}
              value={age?.toString() ?? ""}
              onChange={(c) => {
                const val = parseInt(c.currentTarget.value, 10);
                if (!val || val < 0) setAge(undefined);
                else setAge(val);
              }}
              variant="filled"
            />
          )}
        </Grid>
        <Grid item sm={12} md={4}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControl className={sharedClasses.formItem} variant="filled">
              <InputLabel htmlFor="gender" id="gender-label">
                Gender
              </InputLabel>
              <Select
                labelId="gender-label"
                id="gender"
                value={gender}
                onChange={(c) =>
                  setGender(c.target.value as "Male" | "Female" | undefined)
                }
                native
              >
                <option value={undefined}>Not Set</option>
                <option value="Female">Female</option>
                <option value="Male">Male</option>
              </Select>
            </FormControl>
          )}
        </Grid>
        <Grid item sm={12} md={4}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControl className={sharedClasses.formItem} variant="filled">
              <InputLabel htmlFor="category" id="category-label">
                Category
              </InputLabel>
              <Select
                id="category"
                labelId="category-label"
                value={category}
                onChange={(c) => {
                  if (typeof c.target.value === "number")
                    setCategory(c.target.value as number);
                  else setCategory(parseInt(c.target.value as string, 10));
                }}
                native
              >
                <option value={1}>Cat 1</option>
                <option value={2}>Cat 2</option>
                <option value={3}>Cat 3</option>
                <option value={4}>Cat 4</option>
                <option value={5}>Cat 5</option>
              </Select>
            </FormControl>
          )}
        </Grid>
      </Grid>
      <Grid container spacing={2}>
        <Grid item sm={12} md={6}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControlLabel
              control={
                <Switch
                  checked={drove}
                  onChange={(c) => setDrove(c.currentTarget.checked)}
                />
              }
              label="Drove to Patient"
            />
          )}
        </Grid>
        <Grid item sm={12} md={6}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControlLabel
              control={
                <Switch
                  checked={blueLights}
                  onChange={(c) => setBlueLights(c.currentTarget.checked)}
                />
              }
              label="Used Blue Lights"
            />
          )}
        </Grid>
        <Grid item xs={12}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <TextField
              label="Notes"
              multiline
              className={sharedClasses.formItem}
              value={notes}
              onChange={(c) => {
                setNotes(c.currentTarget.value);
              }}
              variant="filled"
              rows={6}
              id="notes"
            />
          )}
        </Grid>
        <Grid item sm={12} md={6}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControl
              className={sharedClasses.formItem}
              required
              variant="filled"
            >
              <InputLabel htmlFor="outcome" id="outcome-label">
                Outcome
              </InputLabel>
              <Select
                id="outcome"
                labelId="outcome-label"
                value={outcome}
                onChange={(c) => setOutcome(c.target.value as Outcome)}
                native
              >
                <option value="StoodDown">Stood Down</option>
                <option value="Conveyed">Conveyed</option>
                <option value="DischargedOnScene">Discharge on Scene</option>
                <option value="NotFound">Not Found</option>
                <option value="Other">Other</option>
              </Select>
            </FormControl>
          )}
        </Grid>
        <Grid item sm={12} md={6}>
          {isLoading ? (
            <Skeleton className={sharedClasses.formItemSkeleton} />
          ) : (
            <FormControlLabel
              control={
                <Switch
                  checked={reflectionFlag}
                  onChange={(c) => setReflectionFlag(c.currentTarget.checked)}
                />
              }
              label="Mark for Reflection"
            />
          )}
        </Grid>
        <Grid container item xs={12} justify="flex-end">
          <Grid item>
            <Button
              startIcon={<SaveIcon />}
              color="primary"
              variant="contained"
              onClick={submit}
              disabled={!canSubmit}
            >
              Save
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </form>
  );
}
