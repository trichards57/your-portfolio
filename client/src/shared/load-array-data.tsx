import { useAuth0 } from "@auth0/auth0-react";
import { useCallback, useEffect, useState } from "react";
import { useHistory } from "react-router-dom";

function useLoadedArrayData<T extends { id: string }>(
  uri: string,
  deleteUri?: string
) {
  const [errorLoading, setErrorLoading] = useState(false);
  const [errorDeleting, setErrorDeleting] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [isDeleting, setIsDeleting] = useState(false);
  const [data, setData] = useState<T[] | undefined>(undefined);
  const [showUndelete, setShowUndelete] = useState<string[]>([]);
  const { getAccessTokenSilently } = useAuth0();
  const history = useHistory();
  const [totalItems, setTotalItems] = useState<number | undefined>(0);

  useEffect(() => {
    setIsLoading(true);
    setErrorLoading(false);

    const abortController = new AbortController();

    async function loadData() {
      const token = await getAccessTokenSilently({
        audience: "https://tr-toolbox.me.uk/your-portfolio",
      });

      const response = await fetch(uri, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
        signal: abortController.signal,
      });

      if (abortController.signal.aborted) return;

      if (!response.ok) {
        if (response.status === 401) history.push("/");
        else if (response.status === 404) history.push("/home");
        setErrorLoading(true);
      } else {
        const result = (await response.json()) as T[];
        const totalCount = response.headers.get("x-total-count");
        if (totalCount) setTotalItems(parseInt(totalCount, 10));
        else setTotalItems(undefined);
        setData(result);
      }

      setIsLoading(false);
    }
    loadData();
    return () => abortController.abort();
  }, [getAccessTokenSilently, history, uri]);

  function removeUndelete(id: string) {
    setShowUndelete((d) => d.filter((i) => i !== id));
  }

  const deleteItem = useCallback(
    (id: string) => {
      setIsDeleting(true);
      setErrorDeleting(false);

      const abortController = new AbortController();

      async function runItem() {
        const delUri = `${deleteUri}?id=${id}`;

        const token = await getAccessTokenSilently({
          audience: "https://tr-toolbox.me.uk/your-portfolio",
        });

        const response = await fetch(delUri, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
          method: "POST",
          signal: abortController.signal,
        });

        if (abortController.signal.aborted) return;

        if (!response.ok) {
          if (response.status === 401) history.push("/");
          else if (response.status === 404) history.push("/home");
          setErrorDeleting(true);
        } else {
          setData((d) => d && d.filter((i) => i.id !== id));
          setTotalItems((s) => (s === undefined ? undefined : s - 1));
          setShowUndelete((i) => [...i, id]);
        }
        setIsDeleting(false);
      }

      runItem();
      return () => abortController.abort();
    },
    [getAccessTokenSilently, history, deleteUri]
  );

  return {
    data,
    isDeleting,
    isLoading,
    errorDeleting,
    errorLoading,
    showUndelete,
    totalItems,
    deleteItem,
    removeUndelete,
  };
}

export default useLoadedArrayData;
