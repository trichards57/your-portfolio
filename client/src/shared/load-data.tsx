import { useAuth0 } from "@auth0/auth0-react";
import { useEffect, useState } from "react";
import { useHistory } from "react-router-dom";

function useLoadedData<T>(uri: string) {
  const [errorLoading, setErrorLoading] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [data, setData] = useState<T | undefined>(undefined);
  const { getAccessTokenSilently } = useAuth0();
  const history = useHistory();

  useEffect(() => {
    setIsLoading(true);
    setErrorLoading(false);

    const abortController = new AbortController();

    async function loadData() {
      const token = await getAccessTokenSilently({
        audience: "https://tr-toolbox.me.uk/your-portfolio",
      });

      const response = await fetch(uri, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
        signal: abortController.signal,
      });

      if (abortController.signal.aborted) return;

      if (!response.ok) {
        if (response.status === 401) history.push("/");
        else if (response.status === 404) history.push("/home");
        setErrorLoading(true);
      } else {
        const result = (await response.json()) as T;
        setData(result);
      }

      setIsLoading(false);
    }
    loadData();
    return () => abortController.abort();
  }, [getAccessTokenSilently, history, uri]);

  return {
    data,
    isLoading,
    errorLoading,
  };
}

export default useLoadedData;
