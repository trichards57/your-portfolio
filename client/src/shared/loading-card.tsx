import { Card, CardContent, Typography, makeStyles } from "@material-ui/core";
import React from "react";
import { Skeleton } from "@material-ui/lab";

const useStyles = makeStyles({
  date: {
    fontSize: 14,
    width: "50%",
  },
  trimText: {
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
    overflow: "hidden",
  },
  location: {
    marginBottom: 12,
  },
});

function ShiftCard() {
  const classes = useStyles();

  return (
    <Card>
      <CardContent>
        <Typography className={classes.date} gutterBottom>
          <Skeleton />
        </Typography>
        <Typography className={classes.trimText} variant="h5" component="h2">
          <Skeleton />
        </Typography>
        <Typography className={classes.location} color="textSecondary">
          <Skeleton />
        </Typography>
        <Typography variant="body2" component="p">
          <Skeleton />
        </Typography>
        <Typography variant="body2" component="p">
          <Skeleton />
        </Typography>
      </CardContent>
    </Card>
  );
}

export default ShiftCard;
