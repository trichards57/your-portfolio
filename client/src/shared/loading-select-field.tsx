import {
  SelectProps,
  Select,
  InputLabel,
  FormControl,
} from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import React from "react";
import useSharedStyles from "./shared-styles";

function LoadingSelectField(props: SelectProps & { isLoading?: boolean }) {
  const {
    isLoading,
    required,
    variant,
    id,
    labelId,
    children,
    label,
    ...rest
  } = props;
  const sharedClasses = useSharedStyles();

  if (isLoading) return <Skeleton className={sharedClasses.formItemSkeleton} />;

  return (
    <FormControl
      className={sharedClasses.formItem}
      required={required}
      variant={variant}
    >
      <InputLabel htmlFor={id} id={labelId}>
        {label}
      </InputLabel>
      <Select id={id} labelId={labelId} {...rest}>
        {children}
      </Select>
    </FormControl>
  );
}

export default LoadingSelectField;
