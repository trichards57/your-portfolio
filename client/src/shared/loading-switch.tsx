import { FormControlLabel, Switch, SwitchProps } from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import React from "react";
import useSharedStyles from "./shared-styles";

function LoadingSwitch(
  props: SwitchProps & { isLoading?: boolean; label: string }
) {
  const { isLoading, label, ...rest } = props;
  const sharedClasses = useSharedStyles();

  if (isLoading) return <Skeleton className={sharedClasses.formItemSkeleton} />;

  return <FormControlLabel control={<Switch {...rest} />} label={label} />;
}

export default LoadingSwitch;
