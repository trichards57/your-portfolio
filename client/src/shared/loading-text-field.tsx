import { TextField, TextFieldProps } from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import React from "react";
import useSharedStyles from "./shared-styles";

function LoadingTextField(props: TextFieldProps & { isLoading?: boolean }) {
  const { isLoading, ...rest } = props;
  const sharedClasses = useSharedStyles();

  if (isLoading) return <Skeleton className={sharedClasses.formItemSkeleton} />;

  return <TextField {...rest} />;
}

export default LoadingTextField;
