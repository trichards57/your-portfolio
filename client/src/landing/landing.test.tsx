import ReactDOM from "react-dom";
import Landing from ".";
import renderer from "react-test-renderer";
import { useAuth0 } from "@auth0/auth0-react";
import { ServerAudience } from "../shared/constants";
import { fireEvent, render, screen } from "@testing-library/react";

jest.mock("@auth0/auth0-react");

it("renders without crashing", () => {
  const div = document.createElement("div");

  (useAuth0 as jest.Mock).mockReturnValue({
    loginWithRedirect: jest.fn(),
  });

  ReactDOM.render(<Landing />, div);

  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  (useAuth0 as jest.Mock).mockReturnValue({
    loginWithRedirect: jest.fn(),
  });

  const tree = renderer.create(<Landing />).toJSON();

  expect(tree).toMatchSnapshot();
});

it("calls auth0 when login clicked", () => {
  const mockAuth = {
    loginWithRedirect: jest.fn(),
  };
  (useAuth0 as jest.Mock).mockReturnValue(mockAuth);

  render(<Landing />);

  fireEvent.click(screen.getByText("Login"));

  expect(mockAuth.loginWithRedirect).toBeCalledWith({
    audience: ServerAudience,
  });
});

it("calls auth0 when register is clicked and prompts auth0 to register", () => {
  const mockAuth = {
    loginWithRedirect: jest.fn(),
  };
  (useAuth0 as jest.Mock).mockReturnValue(mockAuth);

  render(<Landing />);

  fireEvent.click(screen.getByText("Register"));

  expect(mockAuth.loginWithRedirect).toBeCalledWith({
    audience: ServerAudience,
    screen_hint: "signup",
  });
});
