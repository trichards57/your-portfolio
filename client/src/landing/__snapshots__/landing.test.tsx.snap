// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="MuiContainer-root makeStyles-container-4 MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root makeStyles-paper-5 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <h1
      className="MuiTypography-root MuiTypography-h5"
    >
      Your Portfolio
    </h1>
    <p
      className="MuiTypography-root makeStyles-summary-6 MuiTypography-body1"
    >
      Track your portfolio as you work.
    </p>
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Login
          </span>
        </button>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Register
          </span>
        </button>
      </div>
    </div>
  </div>
</main>
`;
